stream_alert/alert_processor/outputs/credentials/provider.py

```diff
@@ -0,0 +1,589 @@
+"""
+Copyright 2017-present, Airbnb Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import json
+import os
+import shutil
+import tempfile
+from abc import abstractmethod
+
+import boto3
+from botocore.exceptions import ClientError
+
+from stream_alert_cli.outputs.helpers import kms_encrypt, send_creds_to_s3
+
+from stream_alert.shared.logger import get_logger
+
+
+LOGGER = get_logger(__name__)
+
+
+class OutputCredentialsProvider(object):
+    """OutputCredentialsProvider is a helper service to OutputDispatcher that helps it load
+       credentials that are housed on AWS S3, or cached locally.
+
+       OutputDispatcher implementations may require credentials to authenticate with an external
+       gateway. All credentials for OutputDispatchers are to be stored in a single bucket on AWS S3
+       and are encrypted with AWS KMS. When the OutputDispatchers are booted, this these encrypted
+       credentials are downloaded and cached locally on the filesystem. Then, AWS KMS is used to
+       decrypt the credentials when in use.
+
+    Public methods:
+        load_credentials: Returns a dict of the credentials requested
+        get_local_credentials_temp_dir(): Returns full path to a temporary directory where all
+            encrypted credentials are cached.
+
+    """
+
+    def __init__(self, config, defaults, region, service_name, prefix=None, aws_account_id=None):
+        self._config = config
+        self._region = region
```

Comment:
subclasses of `OutputCredentialsProvider` that use 'boto3' clients and need a region could use the `default_config` function to return a botocore config (see: https://github.com/airbnb/streamalert/blob/master/stream_alert/shared/helpers/boto.py#L28)

TBH, I'd advocate for using the `default_config` method anywhere we are creating boto3 clients, since it enforces a shorter timeout (which is better for efficiency/cost in lambda).

Reply:
👍 
So we would do something like

```
client = boto3.client('zzz', config=default_config(region=self._region))
```

Instead of
```
client = boto3.client('zzz', region_name=self._region)
```
?

Reply:
no you could omit region entirely from the call to `default_config` (unless it could/would ever be something other than the current region.. which I don't think it would be in this case)

Reply:
`client = boto3.client('zzz', config=default_config())`

Reply:
I remembered something when I was confirming whether or not region can be globally configured; what about on the [CLI handler](https://github.com/airbnb/streamalert/pull/878/files#diff-21436558def6ebba39502261900e910f)? It looks like the `region` (among other configurations) is passed in from ... somewhere. How does the CLI handler know which `REGION` the code runs in? Is it inferred from the `aws2fa.sh` session?